Dockerfile-alpine.template

```diff
@@ -75,11 +75,10 @@ RUN set -ex \
 	&& make install \
 	\
 	&& runDeps="$( \
-		scanelf --needed --nobanner --recursive /usr/local \
-			| awk '{ gsub(/,/, "\nso:", $2); print "so:" $2 }' \
-			| sort -u \
-			| xargs -r apk info --installed \
+		scanelf --needed --nobanner --format '%n#p' --recursive /usr/local \
+			| tr ',' '\n' \
 			| sort -u \
+			| awk 'system("[ -e /usr/local/lib/" $1 " ]") == 0 { next } { print "so:" $1 }' \
```

Comment:
May I know what's the purpose of the spaces surrounding `$1`?

```awk
system("[ -e /usr/local/lib/" $1 " ]")
```

Anyway, shouldn't we do

```awk
system("[ -e /usr/local/lib/**/" $1 " ]")
```

to match libraries in subdirectories too?

Reply:
> May I know what's the purpose of the spaces surrounding `$1`?

Just syntax -- makes it more readable.  `awk` will concatenate the strings directly anyhow.

> to match libraries in subdirectories too?

No, because `[ -e /usr/local/lib/**/libxyz.so ]` doesn't work properly in a shell (it could partially work if we enabled `globstar` in the subshell that `awk` spawns, but that's a bit heavy, and will still break completely if there's more than one result).  If we needed that sort of check, we'd have to make that string a lot more complex.

In the case of Ruby, we know it's only installing `.so` files directly in `/usr/local/lib`, so those are all we need to filter out (so that we don't accidentally keep packages that provide `.so` files with the same basename, since `apk`'s `.so` dependencies work based on the file basenames).

Reply:
I'm asking because it's relevant for https://github.com/docker-library/official-images/pull/4404, specifically for Varnish 4.1 :smile:

I did not enable the `globstar` option but it seems to work? (For both `/usr/local/lib/*.so` in the case of Varnish 6.0 and `/usr/local/lib/varnish/*.so` in the case of Varnish 4.1)

Reply:
Huh, that's strange -- I would imagine that without `globstar` enabled it's acting just like a single glob, but then it wouldn't match `/usr/local/lib/xxx.so`, so that's interesting.  I wonder what adding `set -x; [ ...` ala `system("set -x; [ -e /usr/local/lib/" $1 " ]")` would show (since that should show the exact path that's getting matched).

Reply:
Hmm... Can't say I understand what's going on:

```
+ scanelf --needed --nobanner+  --format '%n#p' --recursivetr /usr/local ,
 '\n'
+ sort -u
+ awk 'system("set -x; [ -e /usr/local/lib/**/" $1 " ]") == 0 { next } { print "so:" $1 }'
+ '[' -e '/usr/local/lib/**/libc.musl-x86_64.so.1' ]
+ '[' -e '/usr/local/lib/**/libedit.so.0' ]
+ '[' -e '/usr/local/lib/**/libexecinfo.so.1' ]
+ '[' -e '/usr/local/lib/**/libncursesw.so.6' ]
+ '[' -e '/usr/local/lib/**/libpcre.so.1' ]
+ '[' -e /usr/local/lib/varnish/libvarnish.so ]
+ '[' -e '/usr/local/lib/**/libvarnishapi.so.1' ]
+ '[' -e /usr/local/lib/varnish/libvcc.so ]
+ '[' -e /usr/local/lib/varnish/libvgz.so ]
+ runDeps='so:libc.musl-x86_64.so.1
so:libedit.so.0
so:libexecinfo.so.1
so:libncursesw.so.6
so:libpcre.so.1
so:libvarnishapi.so.1'
```